universal-application-tool-0.0.1/app/services/program/ProgramService.java

```diff
@@ -1,17 +1,88 @@
 package services.program;
 
 import com.google.common.collect.ImmutableList;
+import java.util.Optional;
 
-/** Operations you can perform on Programs. */
+/** Operations you can perform on {@link ProgramDefinition}s. */
 public interface ProgramService {
 
-    /** List all programs. */
-    public ImmutableList<ProgramDefinition> listProgramDefinitions();
+  /**
+   * List all programs.
+   *
+   * @return a list of {@link ProgramDefinition}s
+   */
+  ImmutableList<ProgramDefinition> listProgramDefinitions();
 
-    /** Get the definition for a given program. */
-    public ProgramDefinition getProgramDefinition(String id);
+  /**
+   * Get the definition for a given program.
+   *
+   * @param id the ID of the program to retrieve
+   * @return an {@link Optional} of the {@link ProgramDefinition} for the given ID if it exists, or
+   *     {@code Optional.empty()} if not
+   */
+  Optional<ProgramDefinition> getProgramDefinition(long id);
 
-    /** Create a new program. */
-    public ProgramDefinition createProgram(String name, String description);
+  /**
+   * Create a new program.
+   *
+   * @param name a name for this program
+   * @param description the description of what the program provides
+   * @return the {@link ProgramDefinition} that was created
+   */
+  ProgramDefinition createProgram(String name, String description);
 
+  /**
+   * Adds a {@link BlockDefinition} to the given program.
+   *
+   * @param programId the ID of the program to update
+   * @param blockName a name for the block to add
+   * @param blockDescription a description of what the questions in the block address
+   * @return the updated {@link ProgramDefinition}
+   */
+  ProgramDefinition addBlockToProgram(long programId, String blockName, String blockDescription);
+
+  /**
+   * Insert a new {@link BlockDefinition} at a specific index for a given program.
+   *
+   * @param programId the ID of the program to update
+   * @param blockIndex the index in the block list for the new block
+   * @param blockName a name for the block to add
+   * @param blockDescription a description of what the questions in the block address
+   * @return the updated {@link ProgramDefinition}
+   */
+  ProgramDefinition addBlockToProgram(
+      long programId, int blockIndex, String blockName, String blockDescription);
+
+  /**
+   * Update a {@link BlockDefinition} with a set of questions.
+   *
+   * @param programId the ID of the program to update
+   * @param blockIndex the index of the block to update
+   * @param questionIds an {@link ImmutableList} of questions for the block
+   * @return the updated {@link ProgramDefinition}
+   */
+  ProgramDefinition setBlockQuestions(
+      long programId, int blockIndex, ImmutableList<String> questionIds);
```

Comment:
Suggestion (blocking): rename `questionIds` to `questionDefinitionIds`
Suggestion (non-blocking): check if qd IDs are strings or longs

Reply:
Nice catch! Looks like qd IDs are Strings

Reply:
I wonder if qd IDs should possibly be `long`s instead, based on our repo code. See [this test](https://github.com/seattle-uat/universal-application-tool/blob/main/universal-application-tool-0.0.1/test/repository/QuestionRepositoryTest.java#L29) that references a question ID.